polymorphic.ts

```diff
@@ -0,0 +1,142 @@
+import { JsonValue, Serializable } from "./serializable.ts";
+
+const POLYMORPHIC_RESOLVER_MAP = new Map<
+  unknown,
+  (input: string | JsonValue | Object) => Serializable
+>();
+
+export function registerPolymorphicResolver(
+  classPrototype: unknown,
+  resolver: (input: string | JsonValue | Object) => Serializable,
+): void {
+  POLYMORPHIC_RESOLVER_MAP.set(classPrototype, resolver);
+}
+
+// Parent -> key -> value -> initializer
+const POLYMORPHIC_SWITCH_MAP = new Map<
+  unknown,
+  Map<string | symbol, Map<unknown, () => Serializable>>
+>();
+
+export function registerPolymorphicSwitch(
+  parentPrototype: unknown,
+  propertyName: string | symbol,
+  propertyValue: unknown,
+  initializerFunction: () => Serializable,
+): void {
+  // Assert parentPrototype != null
+
+  let classMap = POLYMORPHIC_SWITCH_MAP.get(parentPrototype);
+  if (!classMap) {
+    POLYMORPHIC_SWITCH_MAP.set(parentPrototype, new Map());
+    classMap = POLYMORPHIC_SWITCH_MAP.get(parentPrototype);
+  }
```

Comment:
I've seen other code like this, I guess using `.has()` with the `if` is counter-intuitive for this scenario?

Reply:
It's slightly more efficient when getting a value that exists as we only have to perform one lookup.

For adding it may be slightly slower if actually loading the value associated with the key takes any time

Reply:
souondds good ty!